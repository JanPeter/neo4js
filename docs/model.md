# Model

The model creates the basic CRUD operations for you. With custom Model classes you can add hooks to the model, e.g. if you want to hash the password of a user before creating one, you can create a `beforeCreate` method on your custom Model as explained in the [hooks](#hooks) section.

## Model methods

* [`create(props)`](#create)
* [`findByGuid(guid)`](#findbyguid)
* [`delete(props)`](#delete)
* [`find(props)`](#find)
* [`findOne(props)`](#findOne)
* [`update(props, newProps)`](#update)

## Hook methods

* [`beforeCreate(props)`](#beforecreate)
* [`afterCreate(instance)`](#aftercreate)
* [`beforeFind(props)`](#beforefind)
* [`afterFind(instance)`](#afterfind)
* [`beforeUpdate(props, newProps)`](#beforeupdate)
* [`afterUpdate(instance)`](#afterupdate)

## `Model<Properties, ModelInstance<Properties>>(label: string)`

To create a new Model you need to create a new instance of Model with a given label. The label is used for the neo4j database. If you are not familiar with the concept of a label in neo4j, it's like the name of a table in a relational database. For better flow support you can add the type for the Properties and the ModelInstance with the Properties of your model.

### Example

```js
import { Model, ModelInstance } from "neo4-js";

type TaskProps = {
    title?: string,
    done?: boolean,
};

const Task: Model<TaskProps, ModelInstance<TaskProps>> = new Model("Task");

```

## `create(props: Properties): Promise<ModelInstance<Properties>>` {#create}

Creates a new instance with the given properties. The properties need to be an object consisting only of primitive types. The `create` function automatically generates an uuidv4 and appends it to the properties of your instance via the `guid` property. It returns a Promise of a `ModelInstance` with your properties.

### Example

```js
import { Model, ModelInstance } from "neo4-js";

type TaskProps = {
    title?: StringProperty,
    done?: boolean,
};

const Task = new Model<TaskProps, ModelInstance<TaskProps>>("Task");
Task.create({ title: "Learn how to drive a bike", done: false })
    .then((task: ModelInstance<TaskProps>) => {
        ...
    });
```

## `findByGuid(guid: string): Promise<ModelInstance<Properties>>` {#findbyguid}

Searches for an entity with the given guid. But only for the specific _label_. It returns a Promise which resolves an ModelInstance.

### Example

```js
import { Model, ModelInstance } from "neo4-js";

type TaskProps = {
    title?: StringProperty,
    done?: boolean,
};

const Task = new Model<TaskProps, ModelInstance<TaskProps>>("Task");
Task.create({ title: "Learn how to drive a bike", done: false })
    .then((task: ModelInstance<TaskProps>) => {
        return Task.findByGuid(task.props.guid);
    })
    .then((task: ModelInstance<TaskProps>) => {
        ...
    });
```

## `delete(props: Properties, detach: boolean = false): Promise<number>` {#delete}

Deletes all entities which matches the given Properties. If you set detach to true, it will delete all entities with it's relations. If this is set to false and you want to delete a entity which has a relation to another entity, neo4j will throw an error. It returns a Promise which resolves the number of deleted entities.

### Example

```js
import { Model, ModelInstance } from "neo4-js";

type TaskProps = {
    title?: StringProperty,
    done?: boolean,
};

const Task = new Model<TaskProps, ModelInstance<TaskProps>>("Task");
Task.create({ title: "Learn how to drive a bike", done: false })
    .then((task: ModelInstance<TaskProps>) => {
        return Task.delete({ guid: task.props.guid });
    })
    .then(numberOfDeletedTasks => {
        ...
    });
Task.delete({ title: { $sw: "Buy" }})
    .then((numberOfDeletedTasks: number) => {
        ...
    });
```

## `find(props: Properties): Promise<Array<ModelInstance<Properties>>>` {#find}

Used to search for entities in the neo4j database. The Properties are typically key value types, whereas the value types are primitives such as string, number or boolean. But to create rich queries you can use the following keywords as a key. It returns a Promise which resolves an array of ModelInstances.

```
String operations:
$sw: string - starts with
$ew: string - ends with
$contains: string - string contains
$reg: string - regular expression

Number operations:
$gt: number - greater than
$gte: number - greater than equal
$lt: number - less than
$lte: number - less than equal
$between: number[] - number between two given numbers

Common:
$eq: string | number | boolean - equals (actually the default, so this is not really a useful operation...)
$in: string[] | number[] - value in list of values
$not: any - not...
$or: any[] - concatenates operations with a logical or
$and: any[] - concatenates operations with a logical and
```

### Example

```js
import { Model, ModelInstance } from "neo4-js";

type TaskProps = {
    title?: StringProperty,
    done?: boolean,
};

const Task = new Model<TaskProps, ModelInstance<TaskProps>>("Task");
Task.find({ title: { $or: [ { $sw: "Buy" }, { $sw: "Learn" } ] } })
    .then((tasks: ModelInstance<TaskProps>[]) => {
        ...
    });
Task.find({ title: "Buy milk" })
    .then((tasks: ModelInstance<TaskProps>[]) => {
        ...
    });
```

## `findOne(props: Properties): Promise<ModelInstance<Properties>>` {#findone}

FindOne works the same way as find, but the Promise resolves only one ModelInstance instead of an array of ModelInstances. For more details please have a look at [`find(props: Properties)`](#find).

## `update(props: Properties, newProps: Properties): Promise<Array<ModelInstance<Properties>>>` {#update}

The update method is used to update all entities which match for the first parameter of properties. The matched entities will then be updated with the newProps properties, which merges with the existing ones.

### Example

```js
import { Model, ModelInstance } from "neo4-js";

type TaskProps = {
    title?: StringProperty,
    done?: boolean,
};

const Task = new Model<TaskProps, ModelInstance<TaskProps>>("Task");
Task.update({ title: { $or: [ { $sw: "Buy" }, { $sw: "Learn" } ] } }, { done: true })
    .then((tasks: ModelInstance<TaskProps>[]) => {
        ...
    });
Task.find({ title: "Buy milk" }, { done: true })
    .then((tasks: ModelInstance<TaskProps>[]) => {
        ...
    });
```

## Hooks

### `beforeCreate(props: Properties): Properties` {#beforecreate}

### `afterCreate(instance: ModelInstance<Properties>): ModelInstance<Properties>` {#aftercreate}

### `beforeFind(props: Properties): Properties` {#beforefind}

### `afterFind(instance: ModelInstance<Properties>): ModelInstance<Properties>` {#afterfind}

### `beforeUpdate(props: Properties, newProps: Properties): Properties` {#beforeupdate}

### `afterUpdate(instance: ModelInstance<Properties>): ModelInstance<Properties>` {#afterupdate}
