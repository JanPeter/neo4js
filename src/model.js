import ModelObject from './model-object';
import Query from './query';
import Utils from './Utils';
import Immutable from 'immutable';

const CharGenerator = Utils.CharGenerator;

export default class Model {
  /*
   * For now this function only checks for the exists() constraint.
   * We need to check this manually because the neo4j exists()
   * function only works with the enterprise version.
   *
   * In future purpose maybe I'll add a type system like Sequelize
   */
  _checkProperties(properties) {
    const invalid = this.properties.filter(p => this.schema[p].exists && !properties[p]);

    if (invalid.length) {
      return new Error(`Exists constraint error (NOT NULL for SQL-people).` +
                       `The following fields are not given: ${invalid.join(', ')}`);
    }
  }

  _addDefaultProperties(properties) {
    const newProperties = { ...properties };
    const defaultProperties = this.properties.filter(p => this.schema[p].defaultValue && !properties[p]);

    for (const p of defaultProperties) {
      newProperties[p] = this.schema[p].defaultValue();
    }
    return newProperties;
  }

  _extractProperties(rawResult) {
    return rawResult.records.map(r => r._fields[0].properties);
  }

  _getRelation(o, rel) {
    const query = new Query(this.neo4js);
    const m = CharGenerator.next();
    const n = CharGenerator.next();

    return query
      .match(m, this.labels, { guid: o.guid })
      .relates(rel.relationLabel)
      .match(n, rel.model.labels)
      .ret(n)
      .execute()
      .then((rawResult) => {
        if (rawResult.signature === 127) {
          throw new Error(rawResult);
        }

        const properties = this._extractProperties(rawResult);
        return properties.map(r => this._createModelObject(r, rel.model));
      });
  }

  _getAllRelations(result) {
    const o = Utils._.cloneDeep(result);
    const relations = {};
    const promises = [];

    for (const rel of this.relations) {
      promises.push(this._getRelation(o, rel)
        .then((result) => {
          Utils._.assign(relations, { [rel.propName]: result });
        }));
    }

    return Promise.all(promises)
      .then(() => {
        Utils._.assign(o, relations);
        return this._createModelObject(o);
      });
  }

  _createModelObject(properties, model) {
    model = model || this;
    const o = new ModelObject();
    o.init(properties, this);
    return o;
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Neo4js} neo4js
   */
  constructor(labels, schema, neo4js) {
    if (!labels.sort) {
      labels = [ labels ];
    }
    this.name = labels.sort().join(',');
    this.labels = labels;
    this.schema = Utils._.cloneDeep(schema);

    this.instanceMethods = this.schema.instanceMethods || {};
    delete schema.instanceMethods;

    this.properties = Object.keys(schema);
    this.neo4js = neo4js;
    this.relations = [];
  }

  /**
   * @returns {Promise}
   */
  sync() {
    return Promise.start()
      .then(() => {
        const cmds = [];
        for (const p of this.properties) {
          const prop = this.schema[p];
          if (prop.index) {
            const query = new Query(this.neo4js);
            cmds.push(query.createIndex(this.labels, p));
          }
          if (prop.unique) {
            const query = new Query(this.neo4js);
            cmds.push(query.createConstraint(this.labels, p, 'unique'));
          }
        }

        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} properties
   * @returns {Promise<ModelObject[]>}
   */
  find(properties) {
    return new Promise((resolve, reject) => {
      const m = CharGenerator.next();
      const query = new Query(this.neo4js);

      query
        .match(m, this.labels, properties)
        .ret(m)
        .execute()
        .then(rawResult => {
          if (rawResult.signature === 127) {
            throw new Error(rawResult);
          }

          let properties = this._extractProperties(rawResult);
          return Promise.all(properties.map(p => this._getAllRelations(p)));
        })
        .then((properties) => {
          resolve(properties.map(r => this._createModelObject(r)));
        })
        .catch(err => {
          reject(err);
        });
    })
  }

  /**
   * @param {Model} model
   * @param {String} propertyName
   * @param {String} relationLabel
   */
  hasMany(model, propName, relationLabel) {
    this.relations.push({
      model,
      propName,
      relationLabel,
      type: 'hasMany',
    });
  }

  /**
   * @param {Model} model
   * @param {String} propertyName
   * @param {String} relationLabel
   */
  belongsTo(model, propName, relationLabel) {
    this.relations.push({
      model,
      propName,
      relationLabel,
      type: 'belongsTo',
    });
  }

  /**
   * @param {Object} properties
   * @returns {Promise<ModelObject>}
   */
  findOne(properties) {
    return new Promise((resolve, reject) => {
      const m = CharGenerator.next();
      const query = new Query(this.neo4js);

      query
        .match(m, this.labels, properties)
        .ret(m)
        .limit(1)
        .execute()
        .then(rawResult => {
          if (rawResult.signature === 127) {
            return reject(rawResult);
          }

          const result = this._extractProperties(rawResult);
          if (result.length > 0) {
            resolve(this._getAllRelations(result[0]));
          } else {
            reject(new Error('No node found'));
          }
        })
        .catch(err => {
          reject(err);
        });
    })
  }

  /**
   * @param {Object} properties
   * @returns {Promise}
   */
  count(properties) {
    return new Promise((resolve, reject) => {
      const m = CharGenerator.next();
      const query = new Query(this.neo4js);

      query
        .match(m, this.labels, properties)
        .ret(`count(${m})`)
        .execute()
        .then(rawResult => {
          resolve(rawResult.records[0]._fields[0].low);
        });
    });
  }

  /**
   * @param {ModelObject} from
   * @param {Object} relation
   * @param {String} relation.relationName
   * @param {Object} relation.properties
   * @param {ModelObject} to
   */
  relate(from, rel, to) {
    return new Promise((resolve, reject) => {
      const a = CharGenerator.next();
      const b = CharGenerator.next();
      const query = new Query(this.neo4js);

      query
        .match(a, this.labels, { guid: from.p.guid })
        .relateRight(rel.relationName, rel.properties)
        .match(b, to.model.labels, { guid: to.p.guid })
        .execute()
        .then(rawResult => {
          if (rawResult.signature === 127) {
            return reject(rawResult);
          }
          return resolve(true);
        })
        .catch(err => {
          reject(err);
        });
    });
  }

  /**
   * @param {Object} properties
   * @returns {Promise<ModelObject>}
   */
  create(properties) {
    return new Promise((resolve, reject) => {
      const errors = this._checkProperties(properties);
      if (errors) reject(errors);

      properties = this._addDefaultProperties(properties);

      const m = CharGenerator.next();
      const query = new Query(this.neo4js);

      query
        .create(m, this.labels, properties)
        .ret(m)
        .execute()
        .then(rawResult => {
          if (rawResult.signature === 127) {
            return reject(rawResult);
          }

          const result = this._extractProperties(rawResult);
          if (result.length > 0) {
            resolve(this._createModelObject(result[0]));
          } else {
            reject(new Error('No user inserted'));
          }
        })
        .catch(err => {
          reject(err);
        });
    });
  }
}
