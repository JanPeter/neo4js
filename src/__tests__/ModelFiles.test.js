// @flow

import trineo, { Model, ModelInstance, hasMany, model, hasOne } from "../index";
import idx from "idx";

import { User, UserInstance, Task, TaskInstance } from "./model";
import type { UserProps, TaskProps } from "./model";

describe("Models in different files", () => {
  beforeAll(() => {
    trineo.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  beforeEach(async () => {
    await User.create({ name: "Olaf" });
    await User.create({ name: "Ignatz" });
    await Task.create({ title: "Learn magic", done: true });
    await Task.create({ title: "Write more test cases", done: false });
    await Task.create({ title: "Write more test cases", done: false });
  });

  afterEach(async () => {
    await trineo.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    trineo.close();
  });

  describe("create", () => {
    it("should create a instances of tasks and relate them to person instance", async () => {
      const paul: UserInstance = await User.create({ name: "Paul" });

      const propsArray: TaskProps[] = [
        {
          title: "Buy milk",
        },
        {
          title: "Buy beer",
          done: false,
        },
      ];

      const tasks: TaskInstance[] = await paul.tasks.create(propsArray);

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });
  });
});
