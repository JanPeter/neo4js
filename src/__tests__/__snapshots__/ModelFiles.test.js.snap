// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Models in different files create should create a instances of tasks and relate them to person instance 1`] = `
Array [
  Object {
    "creator": Object {
      "add": [Function],
      "create": [Function],
      "get": [Function],
      "hasOne": [Function],
      "remove": [Function],
      "update": [Function],
    },
    "props": Object {
      "title": "Buy milk",
    },
  },
  Object {
    "creator": Object {
      "add": [Function],
      "create": [Function],
      "get": [Function],
      "hasOne": [Function],
      "remove": [Function],
      "update": [Function],
    },
    "props": Object {
      "done": false,
      "title": "Buy beer",
    },
  },
]
`;
